### PR TITLE
Models + Operations.

### PR DESCRIPTION
**Pull request by leonplondon**
# Summary
Added required DTOs.
Added required models.
Added controllers for each entity.
Added methods inside controllers to retrieve, delete & create each entity.

# Trello Card
https://trello.com/c/HjJkeP2w/8-modelos-para-el-bot

### CODE REVIEW

File: src/main/java/wolox/reactortraining/controllers/TopicController.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+@RestController
+@RequestMapping("/topics")
+public class TopicController {
+
+    @Autowired
+    private TopicRepository topicRepository;
+
+    @PostMapping
+    public Mono<Void> create(@Valid @RequestBody TopicDto topicDto) {
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/controllers/TopicController.java:33]: could set the status code of an HTTP response

File: src/main/java/wolox/reactortraining/controllers/TopicController.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        return Mono
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/controllers/TopicController.java]: could be simplified, the dto has the same fields as the entity, could directly make the insertion, and the repository returns the Mono<entity> as an answer.

File: src/main/java/wolox/reactortraining/controllers/TopicController.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            .flatMap(topicRepository::insert)
+            .then()
+            .log();
+    }
+
+    @DeleteMapping("/{id}")
+    public Mono<Void> delete(@PathVariable("id") String id) {
+        return topicRepository
+            .findById(id)
+            .switchIfEmpty(error(new TopicNotFound()))
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/controllers/TopicController.java]: for clarity, I would place the switchIfEmpty at the end of all operations.

File: src/main/java/wolox/reactortraining/controllers/UserController.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            .flatMap(userRepository::insert)
+            .then()
+            .log();
+    }
+
+    @DeleteMapping("/{id}")
+    public Mono<Void> delete(@PathVariable("id") String id) {
+        return userRepository
+            .findById(id)
+            .switchIfEmpty(Mono.error(new UserNotFound()))
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/controllers/UserController.java]: for clarity, I would place the switchIfEmpty at the end of all operations.

File: src/main/java/wolox/reactortraining/models/Topic.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Id
+    private String id;
+
+    private String description;
+
+    @DBRef
+    private List<User> users;
+
+    public static Topic fromDto(TopicDto dto) {
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/models/Topic.java]: remove this static method, this conversion from dto to entity make it in the constructor

File: src/main/java/wolox/reactortraining/models/User.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Id
+    private String id;
+
+    private String username;
+
+    @DBRef
+    private List<Topic> topics;
+
+    public static User fromDto(UserDto dto) {
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/models/User.java]: same...

File: src/main/java/wolox/reactortraining/models/User.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        User user = new User();
+        user.username = dto.getUsername();
+
+        return user;
+    }
+}
+
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/models/User.java:33]: add method to add a topic to a user

File: src/main/java/wolox/reactortraining/controllers/TopicController.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    @Autowired
+    private TopicRepository topicRepository;
+
+    @PostMapping
+    public Mono<Void> create(@Valid @RequestBody TopicDto topicDto) {
+        return Mono
+            .just(topicDto)
+            .map(Topic::fromDto)
+            .flatMap(topicRepository::insert)
+            .then()
```
Comment (Jhovannycanaspino) [inline src/main/java/wolox/reactortraining/controllers/TopicController.java:37]: could be simplified, it is not necessary to call flatmap and then, the reactive interface of the repository returns a Mono<entity>.

**Review by Jhovannycanaspino — COMMENTED** [review]
Good job!!, I left you a few comments
it would be good to accompany this pr with unit tests of the repository layer, validating the correct insertion of the entities, especially in the case of adding topics to a user.

File: src/main/java/wolox/reactortraining/controllers/TopicController.java (hunk 8 above)
Comment (leonplondon) [reply src/main/java/wolox/reactortraining/controllers/TopicController.java:37]: Returns a Mono<Topic> with `then` we are trying no to leak model details.

File: src/main/java/wolox/reactortraining/models/User.java (hunk 7 above)
Comment (leonplondon) [reply src/main/java/wolox/reactortraining/models/User.java:33]: I created that method for the next card

File: src/main/java/wolox/reactortraining/controllers/TopicController.java (hunk 8 above)
Comment (yanzacd) [reply src/main/java/wolox/reactortraining/controllers/TopicController.java:37]: Why did you use map for topic and flatMap for insert? there are differences between them, map is for synchronous and flatmap is for asynchronous, why did you decide this?

**Review by leonplondon — COMMENTED** [review]
> Good job!!, I left you a few comments
> it would be good to accompany this pr with unit tests of the repository layer, validating the correct insertion of the entities, especially in the case of adding topics to a user.

I think is better to do this when developing the testing card, which explains how to test our Mono, Flux

**Review by Jhovannycanaspino — APPROVED** [review]

**Review by yanzacd — APPROVED** [review]

**Pull request merged**